# Introducción a la Computación

## Computación y Matemática

La computación puede entenderse como una rama de las matemáticas que estudia cómo resolver problemas mediante procedimientos sistemáticos que pueden ser ejecutados por una computadora.

## Hardware y Software

### Hardware

Es todo lo físico, lo que se puede tocar.

Ejemplos:
- Procesador (CPU)
- Memoria RAM
- Disco
- Placa madre
- Periféricos

### Software

Es todo lo intangible, lo que no se puede tocar.

Ejemplos:
- Algoritmos
- Programas
- Lenguajes de programación
- Sistemas operativos

## Niveles de Abstracción en Computación

La computación puede estudiarse desde distintos niveles.

De abajo hacia arriba
1. Hardware
2. Lenguaje ensamblador
3. Lenguajes de programación
4. Programas y aplicaciones que usamos todos los días

Este enfoque suele ser más difícil didácticamente.


De arriba hacia abajo
1. Programas y aplicaciones
2. Lenguajes de programación
3. Algoritmos
4. Hardware

Este enfoque suele ser más intuitivo para aprender.

## Algoritmo

Un algoritmo es una lista ordenada de pasos para resolver un problema.

Ejemplo simple:
El proceso de sumar números es un algoritmo.

Importante:
- Un algoritmo no necesita computadora.
- Es solo el procedimiento lógico.

## De un Problema a un Programa
Problema -> Algoritmo -> Pseudocódigo -> Lenguaje de programación -> Programa -> Computadora


Explicación
1. Problema: Algo que queremos resolver
2. Algoritmo: Pensamos los pasos para resolverlo
3. Pseudocódigo: Escribir esos pasos de forma estructurada sin usar un lenguaje específico
4. Lenguaje de programación: Traducir el algoritmo a un lenguaje que la computadora entienda (C, Python, Java).
5. Programa: el Algoritmo implementado en un lenguaje.

## Pseudocódigo

El pseudocódigo permite describir algoritmos sin depender de un lenguaje específico. Esto permite que pueda implementarse en distintos lenguajes.

Ejemplo:
leer numero
si numero > 0
    imprimir "positivo"
sino
    imprimir "negativo"


## Python y Pseudocódigo

Python es un lenguaje que se parece mucho al pseudocódigo, por lo que resulta muy claro y fácil de leer.

Por eso se usa mucho para enseñar programación.

## Diagramas de Flujo (DDF)

Otra forma de representar algoritmos es mediante diagramas de flujo.

En estos diagramas:
- Rombo → decisiones (IF)
- Rectángulo → procesos
- Óvalo → inicio o fin

## Pensamiento Algorítmico

Antes de programar, siempre hay que:
1. Pensar el problema sin computadora
2. Diseñar el algoritmo
3. Buscar una solución elegante o eficiente

No se trata solo de que funcione, sino de que sea correcto y eficiente.

## Aritmética de Punto Flotante

Los "float" son números con decimales. Su nombre "punto flotante" es porque el punto decimal puede “moverse”.


Cómo maneja los números la computadora?

El procesador tiene unidades separadas para calcular:
- Enteros (integers)
- Punto flotante (floats)

Es decir:
int + int → int
float + float → float

Por eso, cuando hacemos:

In [1]:
0.1+0.2

0.30000000000000004

Nos devuelve: 0.30000000000000004

Es por esto que muchos números no pueden representarse exactamente en una computadora.

Ejemplo:
π (pi) tiene infinitos decimales → La computadora guarda una aproximación.

Por eso:
Cada operación con floats puede introducir pequeños errores de precisión.

## Diferencia entre Matemática y Programación

En matemática, esta operación:
x = x + 1

Es un absurdo (implicaría que 0 = 1).

En programación, significa asignación, y se interpretaría como que el nuevo valor de x = valor actual de x + 1

## Expresiones e Instrucciones

En programación es importante distinguir entre expresiones e instrucciones.

- Expresión → produce un valor cuando se evalúa.
- Instrucción → es una orden que ejecuta una acción en el programa.